assessment/libs/finderiv.php

```diff
@@ -0,0 +1,537 @@
+<?php
+// A library for financial derivative Functions.  Version 1.0  December 27, 2020
+// Author:  Daniel Brown
+//          University of Colorado Boulder
+
+global $allowedmacros;
+array_push($allowedmacros,"finderiv_payout","finderiv_fwdprice","finderiv_fwdpricediv",
+	"finderiv_fwdcontract",	"finderiv_bsm","finderiv_immdate","finderiv_equityfutdate", 
+	"finderiv_convertrate","finderiv_fairforwardrate","finderiv_fra");
+
+// function finderiv_payout
+// calculates the payout of a portfolio given a terminal value for the 
+// underlying asset. The portfolio can contain
+// European calls, European puts, zero coupon bonds and forward contracts
+function finderiv_payout($asset, $types, $possizes, $strikes){
+	if (!is_numeric($asset) ) {
+		echo "finderiv_payout: invalid asset level";
+		return false;
+	}
+	if (!is_array($types)) {
+		echo 'finderiv_payout: types must be an array';
+		return false;
+	}
+	$numTypes = count($types);
+	if (!is_array($possizes)) {
+		echo 'finderiv_payout: position sizes must be an array';
+		return false;
+	}
+	$numpossizes = count($possizes);
+	if (!is_array($strikes)) {
+		echo 'finderiv_payout: strikes must be an array';
+		return false;
+	}
+	$numStrikes = count($strikes);
+	if ($numTypes!=$numpossizes) {
+		echo 'finderiv_payout: type and position size arrays must be of the same size.';
+		return false;
+	}
+	if ($numTypes!=$numStrikes) {
+		echo 'finderiv_payout: type and strike arrays must be of the same size.';
+		return false;
+	}
+	$total = 0;
+    for ($i=0;$i<$numTypes;$i++) {
```

Comment:
Since you're looping on this, it'd be good to add a reasonable size check above (some max value it can't exceed), to prevent super-big $numTypes values that could cause system-locking loops.

Reply:
I added a maximum number.